### ISSUE TITLE
DDE problem need help！

### ISSUE THREAD

**Issue by liushang0322** [SciML/DifferentialEquations.jl#937]
I am a senior matlab user, I now hope to use Julia to solve DelayDifferentialEquations, here I took out an example of matlab, https://www.mathworks.com/help/releases/R2022a/matlab/math/cardiovascular-model-with-discontinuities.html#CardiovascularDDEExample-2, I have made some attempts but still can't get the same integral route, what is wrong with my use?

```julia
mutable struct P_st
    ca
    cv
    R
    r
    Vstr
    alpha0
    alphas
    alphap
    alphaH
    beta0
    betas
    betap
    betaH
    gammaH
end
pppp = P_st(1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 1)
pppp.ca = 1.55
pppp.cv = 519
pppp.R = 1.05
pppp.r = 0.068
pppp.Vstr = 67.9
pppp.alpha0 = 93
pppp.alphas = 93
pppp.alphap = 93
pppp.alphaH = 0.84
pppp.beta0 = 7
pppp.betas = 7
pppp.betap = 7
pppp.betaH = 1.17
pppp.gammaH = 0
const ppppp = pppp
const tau = 4

function ddefun_3(du, u, h, p, t)
    p = deepcopy(ppppp)
    if t <= 600
        p.R = 1.05
    else
        p.R = 0.21 * exp(600 - t) + 0.84
    end
    Paoft = u[1]
    Pvoft = u[2]
    Hoft = u[3]
    Patau = h(p, t - tau)[1]

    

    du[1] = -(1 / (p.ca * p.R)) * Paoft + (1 / (p.ca * p.R)) * Pvoft + (1 / p.ca) * p.Vstr * Hoft

    du[2] = (1 / (p.cv * p.R)) * Paoft
    -(1 / (p.cv * p.R) + 1 / (p.cv * p.r)) * Pvoft

    Ts = 1 / (1 + (Patau / p.alphas)^p.betas)
    Tp = 1 / (1 + (p.alphap / Paoft)^p.betap)

    du[3] = (p.alphaH * Ts) / (1 + p.gammaH * Tp) - p.betaH * Tp
end

function h(p, t)
    ph = deepcopy(p)
    P0 = 93
    Paval = copy(P0)
    Pvval = (1 / (1 + ph.R / ph.r)) * P0
    Hval = (1 / (ph.R * ph.Vstr)) * (1 / (1 + ph.r / ph.R)) * P0
    return [Paval, Pvval, Hval]
end


tspan = [0 1000.0]

prob = DDEProblem(ddefun_3, h(ppppp, 0), h, tspan)
alg = MethodOfSteps(Vern6())


sol = solve(prob, alg)
```

this is matlab code:
```matlab
p.ca     = 1.55;
p.cv     = 519;
p.R      = 1.05;
p.r      = 0.068;
p.Vstr   = 67.9;
p.alpha0 = 93;
p.alphas = 93;
p.alphap = 93;
p.alphaH = 0.84;
p.beta0  = 7;
p.betas  = 7;
p.betap  = 7;
p.betaH  = 1.17;
p.gammaH = 0;

tau = 4;

P0 = 93;
Paval = P0;
Pvval = (1 / (1 + p.R/p.r)) * P0;
Hval = (1 / (p.R * p.Vstr)) * (1 / (1 + p.r/p.R)) * P0;
history = [Paval; Pvval; Hval];


options = ddeset('Jumps',600);
tspan = [0 1000];
sol = dde23(@(t,y,Z) ddefun(t,y,Z,p), tau, history, tspan, options);

function dydt = ddefun(t,y,Z,p) % equation being solved
    if t <= 600
      p.R = 1.05;
    else
      p.R = 0.21 * exp(600-t) + 0.84;
    end    
    ylag = Z(:,1);
    Patau = ylag(1);
    Paoft = y(1);
    Pvoft = y(2);
    Hoft  = y(3);

    dPadt = - (1 / (p.ca * p.R)) * Paoft ...
            + (1/(p.ca * p.R)) * Pvoft ...
            + (1/p.ca) * p.Vstr * Hoft;

    dPvdt = (1 / (p.cv * p.R)) * Paoft...
            - ( 1 / (p.cv * p.R)...
            + 1 / (p.cv * p.r) ) * Pvoft;

    Ts = 1 / ( 1 + (Patau / p.alphas)^p.betas );
    Tp = 1 / ( 1 + (p.alphap / Paoft)^p.betap );

    dHdt = (p.alphaH * Ts) / (1 + p.gammaH * Tp) ...
           - p.betaH * Tp;

    dydt = [dPadt; dPvdt; dHdt];
end     
```

**Comment by ChrisRackauckas** [comment]
What error did you get? Did you try `tspan = (0,1000.0)`?

`p = deepcopy(ppppp)`: that's really weird. Just pass `p`? Also, using a mutable struct like that is going to be really slow, I don't recommend you do that.

**Comment by liushang0322** [comment]
> 你得到了什么错误？你尝试了吗`tspan = (0,1000.0)`？
> 
> `p = deepcopy(ppppp)`: 这真的很奇怪。只是通过`p`？另外，使用像这样的可变结构会非常慢，我不建议你这样做。

I didn't get an error, but I found that the results of matlab did not correspond to each other. I am curious what went wrong.

**Comment by ChrisRackauckas** [comment]
Did you try reducing the tolerances on both to 1e-12?

**Comment by liushang0322** [comment]
Thanks for your reply, tried your method and found that it didn't change the trend of the results

```julia
sol = solve(prob, alg,abstol = 1e-12,reltol = 1e-12)
```

**Comment by ChrisRackauckas** [comment]
Did you do that to MATLAB too? Did you test that the two functions give exactly the same numerical results?

**Comment by devmotion** [comment]
I had a look at the MATLAB docs and that's a possible Julia code (that could be optimized even more):
```julia
using DelayDiffEq
using Plots

# Cardiovascular DDE model by Ottesen (in-place)
function f!(du, u, h, p, t)
    # Extract components of current state
    Pₐ, Pᵥ, H = u

    # Compute mean arterial pressure with time delay
    Pₐτ = h(p, t - p.τ)[1]

    # Compute current peripheral resistance
    R = p.R∞ + exp(min(0, p.tR - t)) * (p.R₀ - p.R∞)

    # Compute derivatives
    du[1] = dPₐdt = (Pᵥ - Pₐ + R * p.Vₛₜᵣ * H) / (p.cₐ * R)
    du[2] = dPᵥdt = (Pₐ - (1 + R / p.r) * Pᵥ) / (p.cᵥ * R)
    Tₛ = 1 / (1 + (Pₐτ / p.αₛ)^p.βₛ)
    Tₚ = 1 / (1 + (p.αₚ / Pₐ)^p.βₚ)
    du[3] = dHdt = (p.αH * Tₛ) / (1 + p.γH * Tₚ) - p.βH * Tₚ

    nothing
end

# History function of the DDE model (out-of-place)
function h(p, t)
    # Compute current peripheral resistance
    R = p.R∞ + exp(min(0, p.tR - t)) * (p.R₀ - p.R∞)

    # Compute history
    Pₐ = p.P₀
    Pᵥ = p.P₀ / (1 + R / p.r)
    H = p.P₀ / (p.Vₛₜᵣ * (R + p.r))

    return [Pₐ, Pᵥ, H]
end

# Parameters
params = (
    τ = 4, # time delay
    cₐ = 1.55, # arterial compliance
    cᵥ = 519, # venous compliance
    R₀ = 1.05, # initial peripheral resistance
    R∞ = 0.84, # limit of peripheral resistance as t → ∞
    tR = 600, # start time of exponential decay of R
    r = 0.068, # venous outflow resistance 
    Vₛₜᵣ = 67.9, # stroke volume
    P₀ = 93, # typical mean arterial pressure
    α₀ = 93,
    αₛ = 93,
    αₚ = 93,
    αH = 0.84,
    β₀ = 7,
    βₛ = 7,
    βₚ = 7,
    βH = 1.17,
    γH = 0,
)
tspan = (0.0, 1000.0) # time span
u₀ = h(params, 0.0) # state at initial time point t=0

# Construct DDE problem with given parameters and specify the constant delay
prob = DDEProblem(f!, u₀, h, tspan, params; constant_lags=(params.τ,))

# Solve the problem and specify a discontinuity at time point tR
# (actually it's only a discontinuity in lower-order derivative) 
alg = MethodOfSteps(Tsit5()) # MethodOfSteps(Rodas5P())
sol = solve(prob, alg; d_discontinuities = (params.tR,))

# Plot heart rate
plot(sol; idxs=3, ylabel="heart rate", legend=false)
```

The strange thing is that the solution is different from the one in the MATLAB docs already *prior* to the discontinuity (regardless of tolerances, solver, or whether steps are upper bounded by the delay). I'm not sure why that's the case, my initial guess was that I made a mistake when implementing the model or when defining the parameter values. So I compared the code with the MATLAB docs and also the MATLAB code there but I did not find an error. Since I don't have MATLAB installed on my computer, unfortunately I can't check if the figure in the docs can be reproduced with the given MATLAB code.

**Comment by devmotion** [comment]
OK, I noticed I have access to MATLAB online and could run the example there without having to install MATLAB on my computer. Running the code in the documentation produces (unsurprisingly) the figure in the MATLAB docs. However, the solution is completely different if I set the tolerances to e.g. `reltol = 1e-6` and `abstol = 1e-6` (I aborted computations with tolerances of `1e-12` since they seemed to be running forever). More precisely, I changed the final lines of code in the MATLAB code to
```
options = ddeset('Jumps',600,'AbsTol',1e-6,'RelTol',1e-6);
tspan = [0 1000];
sol = dde23(@(t,y,Z) ddefun(t,y,Z,p), tau, history, tspan, options);
x = linspace(0, 1000, 100001);
plot(x, deval(sol, x, 3))
title('Heart Rate for Baroreflex-Feedback Mechanism.')
xlabel(['Time t'])
ylabel('H(t)')
```
and got
![dde](https://user-images.githubusercontent.com/26102/221385185-659bdc09-1088-4506-b775-416af9e8f26f.png)

With the Julia script above I get:
![dde](https://user-images.githubusercontent.com/26102/221385419-c81c6c17-880e-4144-97b3-b56eaf6b0ae2.svg)

Thus without further inspection seems the simulations become quite similar if the tolerances of the MATLAB solver are decreased.
